File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        # iterative approach
        # have a pointer at the beginning of a list, keep the smaller node and keep going
        dummy = node = ListNode()

        while list1 and list2:          # comparing the two head nodes
            if list1.val < list2.val:
                node.next = list1       # smaller node is node.next
                list1 = list1.next      # if this list is chosen, move forward
            else:
                node.next = list2
                list2 = list2.next      # smaller node is node.next
            node = node.next            # if this liust is chosen, move forward

        node.next = list1 or list2

        return dummy.next
        # T(n) = O(n + m)
        # S(n) = O(1)

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        // using a 2 pointer approach so the gap between them is exactly n
        ListNode* dummy = new ListNode(0, head);
        ListNode* left = dummy;
        ListNode* right = head;

        // move right forward n steps
        while(n > 0){
            right = right->next;
            n--;
        }

        // move both pointers until right reaches the end
        while(right != nullptr){
            left = left->next;      // This is the node we want to delete
            right = right->next;
        }

        // skip it by doing left->next->next
        left->next = left->next->next;
        return dummy->next;
    }
};
